# Team “We Tried”
                                                                                            
Ajay Srikar Medidi(U85645387)                  
Bapuji Satyala (U61600601)                                                                                           
Gayatri Sowbhagya Lakshmi Akkireddi (U37764178)                                                                         
Ramyasri Muthineni (U41740753)                                                                                                                                                                                                          
Rehaan Ahmed Abdul (U84763124)                                                                                                                
Sanjana Anumula (U64639710)                                                                                          



## Introduction:


One of the most notable instances of this modern era's industrial globalization is the apparel industry. There are many manual operations in this business, which require a lot of labor. Employee performance in terms of production and delivery in clothing manufacturing enterprises is a major factor in meeting the huge demand for textile products around the world. Hence, monitoring, evaluating, and forecasting the productivity performance of the factory work teams is highly desired by the decision-makers in the apparel business.  

We would like to know what all factors in would influence the productivity of the workers where we consider all factors that would influence and how the stake holders can use this data to evaluate the productivity trend and what all factors, they need to look at for the productivity of their workers to rise to help the company’s profits.


## Plan of Action:

0. Importing Required Libraries
1. Data Exploration and Preprocessing
2. Training and Standardization
3. Modeling
4. Grid and Randomized Search
5. Summary/Analysis

### 0.Importing Required Libraries

In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split,GridSearchCV,RandomizedSearchCV
from sklearn.linear_model import LogisticRegression

from sklearn.impute import SimpleImputer
from sklearn.metrics import mean_absolute_error,mean_squared_error,r2_score,accuracy_score

from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from imblearn.under_sampling import RandomUnderSampler
from imblearn.over_sampling import SMOTE
from sklearn import preprocessing
from imblearn.over_sampling import ADASYN
import warnings
warnings.filterwarnings("ignore")

### Loading the Data

The data is in a CSV file, We'll load the data into a dataframe

In [24]:
df=pd.read_csv("garments_worker_productivity.csv")

In [25]:
df

date   quarter  department        day  team  targeted_productivity  \
0      1/1/2015  Quarter1      sweing   Thursday     8                   0.80   
1      1/1/2015  Quarter1  finishing    Thursday     1                   0.75   
2      1/1/2015  Quarter1      sweing   Thursday    11                   0.80   
3      1/1/2015  Quarter1      sweing   Thursday    12                   0.80   
4      1/1/2015  Quarter1      sweing   Thursday     6                   0.80   
...         ...       ...         ...        ...   ...                    ...   
1192  3/11/2015  Quarter2   finishing  Wednesday    10                   0.75   
1193  3/11/2015  Quarter2   finishing  Wednesday     8                   0.70   
1194  3/11/2015  Quarter2   finishing  Wednesday     7                   0.65   
1195  3/11/2015  Quarter2   finishing  Wednesday     9                   0.75   
1196  3/11/2015  Quarter2   finishing  Wednesday     6                   0.70   

        smv     wip  over_time  incentive  idle_time  idle_men  \
0     26.16  1108.0       7080         98        0.0         0   
1      3.94     NaN        960          0        0.0         0   
2     11.41   968.0       3660         50        0.0         0   
3     11.41   968.0       3660         50        0.0         0   
4     25.90  1170.0       1920         50        0.0         0   
...     ...     ...        ...        ...        ...       ...   
1192   2.90     NaN        960          0        0.0         0   
1193   3.90     NaN        960          0        0.0         0   
1194   3.90     NaN        960          0        0.0         0   
1195   2.90     NaN       1800          0        0.0         0   
1196   2.90     NaN        720          0        0.0         0   

      no_of_style_change  no_of_workers  actual_productivity  
0                      0           59.0             0.940725  
1                      0            8.0             0.886500  
2                      0           30.5             0.800570  
3                      0           30.5             0.800570  
4                      0           56.0             0.800382  
...                  ...            ...                  ...  
1192                   0            8.0             0.628333  
1193                   0            8.0             0.625625  
1194                   0            8.0             0.625625  
1195                   0           15.0             0.505889  
1196                   0            6.0             0.394722  

[1197 rows x 15 columns]

### Exploring Rows and Columns 

In [26]:
df.shape

(1197, 15)

In [27]:
df.describe()

team  targeted_productivity          smv           wip  \
count  1197.000000            1197.000000  1197.000000    691.000000   
mean      6.426901               0.729632    15.062172   1190.465991   
std       3.463963               0.097891    10.943219   1837.455001   
min       1.000000               0.070000     2.900000      7.000000   
25%       3.000000               0.700000     3.940000    774.500000   
50%       6.000000               0.750000    15.260000   1039.000000   
75%       9.000000               0.800000    24.260000   1252.500000   
max      12.000000               0.800000    54.560000  23122.000000   

          over_time    incentive    idle_time     idle_men  \
count   1197.000000  1197.000000  1197.000000  1197.000000   
mean    4567.460317    38.210526     0.730159     0.369256   
std     3348.823563   160.182643    12.709757     3.268987   
min        0.000000     0.000000     0.000000     0.000000   
25%     1440.000000     0.000000     0.000000     0.000000   
50%     3960.000000     0.000000     0.000000     0.000000   
75%     6960.000000    50.000000     0.000000     0.000000   
max    25920.000000  3600.000000   300.000000    45.000000   

       no_of_style_change  no_of_workers  actual_productivity  
count         1197.000000    1197.000000          1197.000000  
mean             0.150376      34.609858             0.735091  
std              0.427848      22.197687             0.174488  
min              0.000000       2.000000             0.233705  
25%              0.000000       9.000000             0.650307  
50%              0.000000      34.000000             0.773333  
75%              0.000000      57.000000             0.850253  
max              2.000000      89.000000             1.120437

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1197 entries, 0 to 1196
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   date                   1197 non-null   object 
 1   quarter                1197 non-null   object 
 2   department             1197 non-null   object 
 3   day                    1197 non-null   object 
 4   team                   1197 non-null   int64  
 5   targeted_productivity  1197 non-null   float64
 6   smv                    1197 non-null   float64
 7   wip                    691 non-null    float64
 8   over_time              1197 non-null   int64  
 9   incentive              1197 non-null   int64  
 10  idle_time              1197 non-null   float64
 11  idle_men               1197 non-null   int64  
 12  no_of_style_change     1197 non-null   int64  
 13  no_of_workers          1197 non-null   float64
 14  actual_productivity    1197 non-null   float64
dtypes: f

In [32]:
df.dtypes

DATE                      object
QUARTER                   object
DEPARTMENT                object
DAY                       object
TEAM                       int64
TARGETED_PRODUCTIVITY    float64
SMV                      float64
WIP                      float64
OVER_TIME                  int64
INCENTIVE                  int64
IDLE_TIME                float64
IDLE_MEN                   int64
NO_OF_STYLE_CHANGE         int64
NO_OF_WORKERS            float64
ACTUAL_PRODUCTIVITY      float64
dtype: object

### Identifying and Handling Any Missing Data

In [29]:
df.isna().sum()

date                       0
quarter                    0
department                 0
day                        0
team                       0
targeted_productivity      0
smv                        0
wip                      506
over_time                  0
incentive                  0
idle_time                  0
idle_men                   0
no_of_style_change         0
no_of_workers              0
actual_productivity        0
dtype: int64

### Renaming Columns

In [30]:
df.columns=df.columns.str.upper()

In [31]:
df.columns

Index(['DATE', 'QUARTER', 'DEPARTMENT', 'DAY', 'TEAM', 'TARGETED_PRODUCTIVITY',
       'SMV', 'WIP', 'OVER_TIME', 'INCENTIVE', 'IDLE_TIME', 'IDLE_MEN',
       'NO_OF_STYLE_CHANGE', 'NO_OF_WORKERS', 'ACTUAL_PRODUCTIVITY'],
      dtype='object')

### Identifying any categorical data in the dataframe and Loading Correctly

In [33]:
uniques=df.nunique()
uniques

DATE                      59
QUARTER                    5
DEPARTMENT                 3
DAY                        6
TEAM                      12
TARGETED_PRODUCTIVITY      9
SMV                       70
WIP                      548
OVER_TIME                143
INCENTIVE                 48
IDLE_TIME                 12
IDLE_MEN                  10
NO_OF_STYLE_CHANGE         3
NO_OF_WORKERS             61
ACTUAL_PRODUCTIVITY      879
dtype: int64

In [34]:
df['QUARTER'].unique()

array(['Quarter1', 'Quarter2', 'Quarter3', 'Quarter4', 'Quarter5'],
      dtype=object)

In [35]:
df['DEPARTMENT'].unique()

array(['sweing', 'finishing ', 'finishing'], dtype=object)

In [36]:
df=df.replace('sweing','sewing')

In [37]:
df['DEPARTMENT'].unique()

array(['sewing', 'finishing ', 'finishing'], dtype=object)

In [38]:
df['DEPARTMENT']=df['DEPARTMENT'].str.strip()

In [39]:
df['DEPARTMENT'].value_counts()

sewing       691
finishing    506
Name: DEPARTMENT, dtype: int64

In [40]:
df['TARGETED_PRODUCTIVITY']=df['TARGETED_PRODUCTIVITY'].round(2)
df['ACTUAL_PRODUCTIVITY']

0       0.940725
1       0.886500
2       0.800570
3       0.800570
4       0.800382
          ...   
1192    0.628333
1193    0.625625
1194    0.625625
1195    0.505889
1196    0.394722
Name: ACTUAL_PRODUCTIVITY, Length: 1197, dtype: float64

In [41]:
df.head(5)

DATE   QUARTER DEPARTMENT       DAY  TEAM  TARGETED_PRODUCTIVITY  \
0  1/1/2015  Quarter1     sewing  Thursday     8                   0.80   
1  1/1/2015  Quarter1  finishing  Thursday     1                   0.75   
2  1/1/2015  Quarter1     sewing  Thursday    11                   0.80   
3  1/1/2015  Quarter1     sewing  Thursday    12                   0.80   
4  1/1/2015  Quarter1     sewing  Thursday     6                   0.80   

     SMV     WIP  OVER_TIME  INCENTIVE  IDLE_TIME  IDLE_MEN  \
0  26.16  1108.0       7080         98        0.0         0   
1   3.94     NaN        960          0        0.0         0   
2  11.41   968.0       3660         50        0.0         0   
3  11.41   968.0       3660         50        0.0         0   
4  25.90  1170.0       1920         50        0.0         0   

   NO_OF_STYLE_CHANGE  NO_OF_WORKERS  ACTUAL_PRODUCTIVITY  
0                   0           59.0             0.940725  
1                   0            8.0             0.886500  
2                   0           30.5             0.800570  
3                   0           30.5             0.800570  
4                   0           56.0             0.800382

In [42]:
df. describe()

TEAM  TARGETED_PRODUCTIVITY          SMV           WIP  \
count  1197.000000            1197.000000  1197.000000    691.000000   
mean      6.426901               0.729632    15.062172   1190.465991   
std       3.463963               0.097891    10.943219   1837.455001   
min       1.000000               0.070000     2.900000      7.000000   
25%       3.000000               0.700000     3.940000    774.500000   
50%       6.000000               0.750000    15.260000   1039.000000   
75%       9.000000               0.800000    24.260000   1252.500000   
max      12.000000               0.800000    54.560000  23122.000000   

          OVER_TIME    INCENTIVE    IDLE_TIME     IDLE_MEN  \
count   1197.000000  1197.000000  1197.000000  1197.000000   
mean    4567.460317    38.210526     0.730159     0.369256   
std     3348.823563   160.182643    12.709757     3.268987   
min        0.000000     0.000000     0.000000     0.000000   
25%     1440.000000     0.000000     0.000000     0.000000   
50%     3960.000000     0.000000     0.000000     0.000000   
75%     6960.000000    50.000000     0.000000     0.000000   
max    25920.000000  3600.000000   300.000000    45.000000   

       NO_OF_STYLE_CHANGE  NO_OF_WORKERS  ACTUAL_PRODUCTIVITY  
count         1197.000000    1197.000000          1197.000000  
mean             0.150376      34.609858             0.735091  
std              0.427848      22.197687             0.174488  
min              0.000000       2.000000             0.233705  
25%              0.000000       9.000000             0.650307  
50%              0.000000      34.000000             0.773333  
75%              0.000000      57.000000             0.850253  
max              2.000000      89.000000             1.120437

### Feature Engineering 

The method of producing new, significant variables from the available data to improve a machine learning model's performance is known as feature engineering. 

In [43]:
df['PRODUCTIVITY']=df['TARGETED_PRODUCTIVITY']-df['ACTUAL_PRODUCTIVITY']

In [44]:
df['PRODUCTIVITY']

0      -0.140725
1      -0.136500
2      -0.000570
3      -0.000570
4      -0.000382
          ...   
1192    0.121667
1193    0.074375
1194    0.024375
1195    0.244111
1196    0.305278
Name: PRODUCTIVITY, Length: 1197, dtype: float64

In [45]:
(df['PRODUCTIVITY']>0).value_counts()

False    875
True     322
Name: PRODUCTIVITY, dtype: int64

In [46]:
df['CLASSIFICATION'] = df['PRODUCTIVITY'].apply(lambda x: 1 if x > 0 else 0)

In [47]:
df['CLASSIFICATION'].value_counts()

0    875
1    322
Name: CLASSIFICATION, dtype: int64

In [48]:
df_dummy=pd.get_dummies(
    df,
    prefix_sep='_',
    dummy_na=False,
    drop_first=True,
    columns=['DEPARTMENT','DAY','QUARTER'],
    dtype='int32'
    
)
df_dummy

DATE  TEAM  TARGETED_PRODUCTIVITY    SMV     WIP  OVER_TIME  \
0      1/1/2015     8                   0.80  26.16  1108.0       7080   
1      1/1/2015     1                   0.75   3.94     NaN        960   
2      1/1/2015    11                   0.80  11.41   968.0       3660   
3      1/1/2015    12                   0.80  11.41   968.0       3660   
4      1/1/2015     6                   0.80  25.90  1170.0       1920   
...         ...   ...                    ...    ...     ...        ...   
1192  3/11/2015    10                   0.75   2.90     NaN        960   
1193  3/11/2015     8                   0.70   3.90     NaN        960   
1194  3/11/2015     7                   0.65   3.90     NaN        960   
1195  3/11/2015     9                   0.75   2.90     NaN       1800   
1196  3/11/2015     6                   0.70   2.90     NaN        720   

      INCENTIVE  IDLE_TIME  IDLE_MEN  NO_OF_STYLE_CHANGE  ...  \
0            98        0.0         0                   0  ...   
1             0        0.0         0                   0  ...   
2            50        0.0         0                   0  ...   
3            50        0.0         0                   0  ...   
4            50        0.0         0                   0  ...   
...         ...        ...       ...                 ...  ...   
1192          0        0.0         0                   0  ...   
1193          0        0.0         0                   0  ...   
1194          0        0.0         0                   0  ...   
1195          0        0.0         0                   0  ...   
1196          0        0.0         0                   0  ...   

      DEPARTMENT_sewing  DAY_Saturday  DAY_Sunday  DAY_Thursday  DAY_Tuesday  \
0                     1             0           0             1            0   
1                     0             0           0             1            0   
2                     1             0           0             1            0   
3                     1             0           0             1            0   
4                     1             0           0             1            0   
...                 ...           ...         ...           ...          ...   
1192                  0             0           0             0            0   
1193                  0             0           0             0            0   
1194                  0             0           0             0            0   
1195                  0             0           0             0            0   
1196                  0             0           0             0            0   

      DAY_Wednesday  QUARTER_Quarter2  QUARTER_Quarter3  QUARTER_Quarter4  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   
...             ...               ...               ...               ...   
1192              1                 1                 0                 0   
1193              1                 1                 0                 0   
1194              1                 1                 0                 0   
1195              1                 1                 0                 0   
1196              1                 1                 0                 0   

      QUARTER_Quarter5  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0  
...                ...  
1192                 0  
1193                 0  
1194                 0  
1195                 0  
1196                 0  

[1197 rows x 24 columns]

In [49]:
#df_dummy=df_dummy.drop(['DATE','SMV','INCENTIVE'],axis=1)

In [50]:
#imputer = SimpleImputer(strategy='mean')

In [51]:
#df_dummy['WIP']=imputer.fit_transform(df_dummy[['WIP']])

In [52]:
#df_dummy = pd.DataFrame(df_imputed, columns=df_dummy.columns)

In [53]:
df_dummy.isna().sum()

DATE                       0
TEAM                       0
TARGETED_PRODUCTIVITY      0
SMV                        0
WIP                      506
OVER_TIME                  0
INCENTIVE                  0
IDLE_TIME                  0
IDLE_MEN                   0
NO_OF_STYLE_CHANGE         0
NO_OF_WORKERS              0
ACTUAL_PRODUCTIVITY        0
PRODUCTIVITY               0
CLASSIFICATION             0
DEPARTMENT_sewing          0
DAY_Saturday               0
DAY_Sunday                 0
DAY_Thursday               0
DAY_Tuesday                0
DAY_Wednesday              0
QUARTER_Quarter2           0
QUARTER_Quarter3           0
QUARTER_Quarter4           0
QUARTER_Quarter5           0
dtype: int64

In [78]:
features=df_dummy.drop(['ACTUAL_PRODUCTIVITY','PRODUCTIVITY','CLASSIFICATION','WIP','DATE'],axis=1)
features
target=df_dummy['CLASSIFICATION']
target

0       0
1       0
2       0
3       0
4       0
       ..
1192    1
1193    1
1194    1
1195    1
1196    1
Name: CLASSIFICATION, Length: 1197, dtype: int64

## Training and Standardization

Training a dataset is nothing but using a part of the data to teach a model how to make predictions, while testing the dataset involves evaluating the model's performance taking a seperate part of the data to assess its performance.

### Training the Dataset

In [79]:
X_train,X_test,y_train,y_test=train_test_split(features,target,test_size=0.2,random_state=1)

In [80]:
X_train

TEAM  TARGETED_PRODUCTIVITY    SMV  OVER_TIME  INCENTIVE  IDLE_TIME  \
994      1                   0.60  26.66       6840         38        0.0   
584      8                   0.70   4.15        960          0        0.0   
824      6                   0.80   2.90       1440          0        0.0   
323      9                   0.80  26.16      10620         63        0.0   
1106     3                   0.80   4.60       1440          0        0.0   
...    ...                    ...    ...        ...        ...        ...   
715      8                   0.35   4.15       1440          0        0.0   
905     12                   0.80  15.26       4080         50        0.0   
1096     9                   0.75  18.79       3480         45        0.0   
235      5                   0.70  20.79      10440          0        0.0   
1061    10                   0.70  21.82       5760         30        0.0   

      IDLE_MEN  NO_OF_STYLE_CHANGE  NO_OF_WORKERS  DEPARTMENT_sewing  \
994          0                   0           57.0                  1   
584          0                   0            8.0                  0   
824          0                   0           12.0                  0   
323          0                   0           59.0                  1   
1106         0                   0           12.0                  0   
...        ...                 ...            ...                ...   
715          0                   0           12.0                  0   
905          0                   0           34.0                  1   
1096         0                   0           51.0                  1   
235          0                   0           58.0                  1   
1061         0                   0           53.0                  1   

      DAY_Saturday  DAY_Sunday  DAY_Thursday  DAY_Tuesday  DAY_Wednesday  \
994              0           1             0            0              0   
584              0           0             0            1              0   
824              0           0             0            0              1   
323              0           0             0            0              0   
1106             0           1             0            0              0   
...            ...         ...           ...          ...            ...   
715              0           0             0            0              1   
905              0           0             0            1              0   
1096             1           0             0            0              0   
235              0           0             0            1              0   
1061             0           0             0            0              1   

      QUARTER_Quarter2  QUARTER_Quarter3  QUARTER_Quarter4  QUARTER_Quarter5  
994                  0                 0                 0                 0  
584                  0                 0                 0                 0  
824                  0                 1                 0                 0  
323                  0                 1                 0                 0  
1106                 1                 0                 0                 0  
...                ...               ...               ...               ...  
715                  1                 0                 0                 0  
905                  0                 0                 1                 0  
1096                 0                 0                 0                 0  
235                  1                 0                 0                 0  
1061                 0                 0                 0                 0  

[957 rows x 19 columns]

### Standardization

Standardization is the process of transforming numerical data to have a mean of 0 and a standard deviation of 1, These are often used to ensure variables are on a same scale for analysis.

In [81]:
scaler = preprocessing.StandardScaler()
scaler.fit(X_train)

# Transform the predictors of training and validation sets
X_train = scaler.transform(X_train) 
X_test = scaler.transform(X_test)

In [82]:
#rus = RandomUnderSampler(random_state=42)
#X_res, y_res = rus.fit_resample(X_train, y_train)

In [83]:
#smote = SMOTE(random_state=42)
#X_res, y_res = smote.fit_resample(X_train, y_train)

In [84]:
adasyn = ADASYN(sampling_strategy='auto', random_state=42, n_neighbors=5, n_jobs=-1)

In [85]:
X_res, y_res = adasyn.fit_resample(X_train, y_train)

In [86]:
y_train

994     0
584     0
824     0
323     0
1106    0
       ..
715     0
905     0
1096    1
235     1
1061    0
Name: CLASSIFICATION, Length: 957, dtype: int64

### Decision Tree

In [87]:
clf=DecisionTreeClassifier()
_=clf.fit(X_res,y_res)

In [88]:
y_pred = clf.predict(X_test) 

In [89]:
ac=accuracy_score(y_test,y_pred)
print("Decision Tree Accuracy:", ac)

Decision Tree Accuracy: 0.8125


### SVC

In [90]:
clf=SVC()
_=clf.fit(X_res,y_res)

In [91]:
y_pred = clf.predict(X_test) 

In [92]:
ac=accuracy_score(y_test,y_pred)
print("SVC Accuracy:", ac)

SVC Accuracy: 0.6875


### Logistic Regression

In [93]:
clf=LogisticRegression()
_=clf.fit(X_res,y_res)

In [94]:
y_pred=clf.predict(X_test)

In [95]:
ac=accuracy_score(y_test,y_pred)
print("Logistic Regression Accuracy:", ac)

Logistic Regression Accuracy: 0.6583333333333333


## Grid and Randomized Search

RandomizedSearchCV combines random search and grid search by randomly sampling hyperparameters from predefined distributions.

### Grid and Randomized Search for Decision Tree

In [97]:
score_measure = 'accuracy'
k_fold = 10
# Considering the Parameters
param_grid = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'max_depth': [int(x) for x in np.linspace(1, 40000, 50)], 
    'max_leaf_nodes': [2, 5, 10, 20, 30, 40],
    'min_samples_split': [2, 5, 10, 20, 40, 60, 80, 100], 
    'min_impurity_decrease': [0, 0.0005, 0.001, 0.005, 0.01], 
}
gridSearch = GridSearchCV(
    DecisionTreeClassifier(random_state=1), 
    param_grid, cv=k_fold, 
    scoring=score_measure,
    n_jobs=-1)  # n_jobs=-1 will utilize all available CPUs 

gridSearch.fit(X_res, y_res)
print(score_measure, 'score: ', gridSearch.best_score_)
print('parameters: ', gridSearch.best_params_)

bestAccuracyTree = gridSearch.best_estimator_

accuracy score:  0.823064628908201
parameters:  {'criterion': 'gini', 'max_depth': 817, 'max_leaf_nodes': 40, 'min_impurity_decrease': 0, 'min_samples_split': 2}


In [98]:
# Defining the hyperparameter search space for Decision Tree
k_fold=10
param_dist = {
     'criterion': ['gini', 'entropy', 'log_loss'],
    'max_leaf_nodes': [None, 5, 10, 20, 30, 40],
    'max_depth': [2, 5, 10, 20, 30, 40],
    'min_samples_split': [2, 5, 10, 20, 40, 60, 80, 100],
    'min_samples_leaf': [1, 2, 4],
    'min_impurity_decrease': [0, 0.0005, 0.001, 0.005, 0.01]
}

# Creating a Decision Tree model
model = DecisionTreeClassifier(random_state=1)

# Initializing RandomizedSearchCV
randomSearch = RandomizedSearchCV(
    model, param_distributions=param_dist, n_iter=10, cv=k_fold, scoring=score_measure, n_jobs=-1, random_state=1)

# Fitting the RandomizedSearchCV to the training data
randomSearch.fit(X_res, y_res)

# Get the best parameters and best estimator
best_params = randomSearch.best_params_
best_estimator = randomSearch.best_estimator_

# Print the best parameters and their corresponding performance
print(score_measure, 'score:', randomSearch.best_score_)
print('Best parameters:', best_params)

bestAccuracyTree = randomSearch.best_estimator_

accuracy score: 0.823064628908201
Best parameters: {'min_samples_split': 5, 'min_samples_leaf': 1, 'min_impurity_decrease': 0.0005, 'max_leaf_nodes': 40, 'max_depth': 40, 'criterion': 'gini'}


### Grid and Randomized Search for LOG GD

In [99]:
score_measure = 'accuracy'
k_fold = 10

# Defining the  hyperparameter search space
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],  # Regularization strength
    'penalty': ['l1', 'l2'],  # Regularization type
    'solver': ['liblinear'],  # Suitable for small datasets
}

# Initializing the  GridSearchCV for the Logistic Regression
gridSearch = GridSearchCV(
    LogisticRegression(random_state=1),
    param_grid,
    cv=k_fold,
    scoring=score_measure,
    n_jobs=-1  # Utilize all available CPUs
)

# Fitting the grid search to  training data
gridSearch.fit(X_res, y_res)

# Getting the best parameters and best estimator
best_params = gridSearch.best_params_
best_estimator = gridSearch.best_estimator_

# Printing the best parameters and best accuracy
print(score_measure, 'score:', gridSearch.best_score_)
print('Best parameters:', best_params)

# storing the best Logistic Regression model:
bestLogisticRegression = gridSearch.best_estimator_

accuracy score: 0.6826740585356108
Best parameters: {'C': 10, 'penalty': 'l1', 'solver': 'liblinear'}


In [110]:
# Defining the hyperparameter search space for Logistic Regression
k_fold=10
param_dist = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'penalty': ['l1', 'l2'],
    'solver': ['liblinear', 'saga'],
    'max_iter': [10000, 200000, 300000, 400000, 500000],
}

# Creating a Logistic Regression model
model = LogisticRegression(random_state=1)

# Initializing RandomizedSearchCV
randomSearch = RandomizedSearchCV(
    model, param_distributions=param_dist, n_iter=10000, cv=k_fold, scoring=score_measure, n_jobs=-1, random_state=1)

# Fitting the RandomizedSearchCV to  training data
randomSearch.fit(X_res, y_res)

# Get the best parameters and best estimator
best_params = randomSearch.best_params_
best_estimator = randomSearch.best_estimator_

# Printing the best parameters and their corresponding performance
print(score_measure, 'score:', randomSearch.best_score_)
print('Best parameters:', best_params)

bestAccuracyLogisticrand = randomSearch.best_estimator_

accuracy score: 0.6826840475476976
Best parameters: {'solver': 'saga', 'penalty': 'l1', 'max_iter': 10000, 'C': 1}


### Grid and Randomized Search for SVM

In [112]:
# Defining the hyperparameter for the SVM 

param_grid = {
    'C': [0.1, 1, 10, 100],  # Regularization parameter
    'kernel': ['rbf','poly','linear'],  # Kernel type
    'degree': [2, 3, 4],  # Degree of the polynomial kernel (if 'poly' is chosen)
    'gamma': ['scale', 'auto'] + [0.001, 0.01, 0.1, 1],  # Kernel coefficient (if 'rbf' or 'poly' is chosen)
}

# Defining the Model (SVC for classification)
model = SVC(random_state=1)



# Initializing the  GridSearchCV for SVM
grid_search = GridSearchCV(
    model, param_grid=param_grid, cv=k_fold, scoring=score_measure, n_jobs=-1
)

# Fitting the GridSearchCV the split data
grid_search.fit(X_res, y_res)

# Getting the best parameters and best estimator
best_params = grid_search.best_params_
best_estimator = grid_search.best_estimator_

# Evaluate the best model on  test data
test_accuracy = best_estimator.score(X_test, y_test)

# Printing the best parameters and test accuracy
print(score_measure, 'score:', grid_search.best_score_)
print('Best Parameters:', best_params)
print('Test Accuracy:', test_accuracy)

accuracy score: 0.8499051043851763
Best Parameters: {'C': 100, 'degree': 2, 'gamma': 'auto', 'kernel': 'rbf'}
Test Accuracy: 0.7041666666666667


In [113]:
# Defining the  hyperparameter search space for SVM
k_fold=10
param_dist = {
    'C': [0.1, 1, 10, 100],
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],
    'gamma': [0.1, 1, 10, 'scale', 'auto'],
    'degree': [2, 3, 4, 5]
}

# Create an SVM classifier
model = SVC()

# Initializing RandomizedSearchCV
randomSearch = RandomizedSearchCV(
    model, param_distributions=param_dist, n_iter=10, cv=k_fold, scoring=score_measure, n_jobs=-1, random_state=1)

# Fitting the RandomizedSearchCV to the training data
randomSearch.fit(X_res, y_res)

# Get the best parameters and best estimator
best_params = randomSearch.best_params_
best_estimator = randomSearch.best_estimator_

# Print the best parameters and their corresponding performance
print(score_measure, 'score:', randomSearch.best_score_)
print('Best parameters:', best_params)

bestAccuracySVM = randomSearch.best_estimator_

accuracy score: 0.8477924283288383
Best parameters: {'kernel': 'rbf', 'gamma': 0.1, 'degree': 2, 'C': 100}


## Summary

We've implemented Decision Tree, SVM and Logistic Regression on our dataset and We've got the following conclusions:

The Accuracies that Each Model Genrated:
Decision Tree:0.8125 
SVM:0.6875
Logistic Regression:0.6583

Grid and Randomized Test Accuracies:

Decision Tree:
accuracy score:  0.823064628908201

SVM:
accuracy score: 0.8477924283288383

Logistic Regression:
accuracy score: 0.6826740585356108

SVM, after hyperparameter optimization, achieved the highest accuracy of 0.8478, making it the best-performing model among the three.

Decision Tree, after hyperparameter tuning, improved slightly but remained the second-best performing model with an accuracy of 0.8231.

Logistic Regression had the lowest accuracy both before and after hyperparameter optimization, indicating that it might not be the most suitable model for this specific task.